src/raftstore/store/worker/region.rs

```diff
@@ -220,62 +193,34 @@ impl Runner {
         let term = apply_state.get_truncated_state().get_term();
         let idx = apply_state.get_truncated_state().get_index();
         let snap_key = SnapKey::new(region_id, term, idx);
-        let snap_file = box_try!(self.mgr.rl().get_snap_file(&snap_key, false));
+        let mut applier = box_try!(self.mgr.rl().get_recv_snapshot_file_applier(&snap_key));
         self.mgr.wl().register(snap_key.clone(), SnapEntry::Applying);
         defer!({
             self.mgr.wl().deregister(&snap_key, &SnapEntry::Applying);
         });
-        if !snap_file.exists() {
-            return Err(box_err!("missing snap file {}", snap_file.path().display()));
+        if !applier.exists() {
+            return Err(box_err!("missing snapshot file {}", applier.path()));
         }
         try!(check_abort(&abort));
-        let mut reader = box_try!(snap_file.reader());
-
         let timer = Instant::now();
         let mut snapshot_size = 0;
         let mut snapshot_kv_count = 0;
-        // Write the snapshot into the region.
-        loop {
-            // TODO: avoid too many allocation
-            try!(check_abort(&abort));
-            let cf = box_try!(reader.decode_compact_bytes());
-            if cf.is_empty() {
-                break;
-            }
-            let handle = box_try!(rocksdb::get_cf_handle(&self.db,
-                                                         unsafe { str::from_utf8_unchecked(&cf) }));
-            let mut wb = WriteBatch::new();
-            let mut batch_size = 0;
-            loop {
-                try!(check_abort(&abort));
-                let key = box_try!(reader.decode_compact_bytes());
-                if key.is_empty() {
-                    box_try!(self.db.write(wb));
-                    snapshot_size += batch_size;
-                    break;
-                }
-                snapshot_kv_count += 1;
-                box_try!(util::check_key_in_region(keys::origin_key(&key), &region));
-                batch_size += key.len();
-                let value = box_try!(reader.decode_compact_bytes());
-                batch_size += value.len();
-                box_try!(wb.put_cf(handle, &key, &value));
-                if batch_size >= self.batch_size {
-                    box_try!(self.db.write(wb));
-                    snapshot_size += batch_size;
-                    wb = WriteBatch::new();
-                    batch_size = 0;
-                }
-            }
+        {
```

Comment:
The ApplyOptions borrows `snapshot_size` and `snapshot_kv_count`, this scope is to end the borrowing.